lib/resourcemerge/core.go

```diff
@@ -202,6 +202,44 @@ func ensureContainerPort(modified *bool, existing *corev1.ContainerPort, require
 	}
 }
 
+func EnsureServicePorts(modified *bool, existing *[]corev1.ServicePort, required []corev1.ServicePort) {
+	for i, existingServicePort := range *existing {
+		var existingCurr *corev1.ServicePort
+		for _, requiredServicePort := range required {
+			if existingServicePort.Name == requiredServicePort.Name {
```

Comment:
Matching by name and then taking the union of in-cluster and in-manifest ports means there's no way to remove an entry.  Also, [`name` is optional][1], although I'm not clear on how that squares [with][2]:

> All ports within a ServiceSpec must have unique names.

Maybe it's "If you set a `name`, it must not duplicate an existing `ports` entry in the same Service"?

Do we even need to support folks adding additional `ports` values to Service manifests managed by the CVO?  Can we just clobber the existing value with the required value and set `modified` if that changed anything?

[1]: https://github.com/kubernetes/api/blob/11707872ac1cab52aefea0325f4698eaffeda3c8/core/v1/types.go#L3985

Reply:
>	// Optional if only one ServicePort is defined on this service.
https://github.com/kubernetes/api/blob/11707872ac1cab52aefea0325f4698eaffeda3c8/core/v1/types.go#L3984

>Matching by name and then taking the union of in-cluster and in-manifest ports means there's no way to remove an entry

There's no union, `required` overwrite `existing` via `ensureServicePort`. The loop below ensures unmatched entries are removed https://github.com/openshift/cluster-version-operator/pull/272/files#diff-f7675c7b39553e431cdc6c3ba8f5c482R221. See test cases for each scenario (with/without name) https://github.com/openshift/cluster-version-operator/pull/272/files#diff-415c13f11ffc32696c5d69b900b3fe58R283.

Instead of a nested loop this could get better performance by hashing the port name but I just wanted to replicate the pattern already present for similar funcs, e.g `ensureContainers` https://github.com/openshift/cluster-version-operator/blob/df38364e842d9a3b5caa2a07be92c1cee2d2651a/lib/resourcemerge/core.go#L64